progress/quantity.go

```diff
@@ -0,0 +1,176 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package progress
+
+import (
+	"fmt"
+	"math"
+)
+
+// these are taken from github.com/chipaca/quantity with permission :-)
+
+func formatAmount(amount uint64, width int) string {
+	if width < 0 {
+		width = 5
+	}
+	max := uint64(5000)
+	maxFloat := 999.5
+
+	if width < 4 {
+		width = 3
+		max = 999
+		maxFloat = 99.5
+	}
+
+	if amount <= max {
+		pad := ""
+		if width > 5 {
+			pad = " "
+		}
+		return fmt.Sprintf("%*d%s", width-len(pad), amount, pad)
+	}
+	var prefix rune
+	r := float64(amount)
+	// zetta and yotta are me being pedantic: maxuint64 is ~18EB
+	for _, prefix = range "kMGTPEZY" {
+		r /= 1000
+		if r < maxFloat {
+			break
+		}
+	}
+
+	width--
+	digits := 3
+	if r < 99.5 {
+		digits--
+		if r < 9.5 {
+			digits--
+			if r < .95 {
+				digits--
+			}
+		}
+	}
+	precision := 0
+	if (width - digits) > 1 {
+		precision = width - digits - 1
+	}
+
+	s := fmt.Sprintf("%*.*f%c", width, precision, r, prefix)
+	if r < .95 {
+		return s[1:]
+	}
+	return s
+}
+
+func formatBPS(n, sec float64, width int) string {
+	if sec < 0 {
+		sec = -sec
+	}
+	return formatAmount(uint64(n/sec), width-2) + "B/s"
+}
+
+const (
+	period = 365.25 // julian years (c.f. the actual orbital period, 365.256363004d)
+)
+
+func divmod(a, b float64) (q, r float64) {
+	q = math.Floor(a / b)
+	return q, a - q*b
+}
+
+func formatDuration(dt float64) string {
```

Comment:
Is there really no nice built-in library function to do all of this?

Reply:
not so fixated on it being fixed-width while still conveying as much info as possible, no